### PR TITLE
Add eastwood lein plugin, enable in CI, fix all warnings.

### PR DESCRIPTION
**Pull request by stevana**

### CODE REVIEW

**Comment by stevana** [comment]
This should help catch problems like the one introduced in my previous PR (sorry about that).

There were a few warnings I wasn't sure how to fix, I marked them with `XXX` in the comments together with the warning they produced. These should be fixed/disabled before merging.

File: jepsen/src/jepsen/generator/interpreter.clj [hunk 1]
```diff
@@ -46,7 +46,8 @@
                      nil
                      (catch Exception e
                        (warn e "Error opening client")
-                       (set (.client this) nil)
```
Comment (stevana) [inline jepsen/src/jepsen/generator/interpreter.clj:49]: Should this be using `set!`?

**Comment by stevana** [comment]
I think the failing test is unrelated to my changes.

File: jepsen/src/jepsen/generator/interpreter.clj (hunk 1 above)
Comment (aphyr) [reply jepsen/src/jepsen/generator/interpreter.clj:49]: Good catch, this is a bug!

**Comment by aphyr** [comment]
You're right--the failing test is a probabilistic test that depends on thread scheduling. I bet you some threads are getting starved when run on Travis-CI.

**Comment by aphyr** [comment]
And--no worries about the compilation errors in the last PR. I don't think there are any tests for jepsen.nemesis.combined, because it was a rush job for a client, and it's side-effect heavy. I should write at least some basic ones--you can't be expected to know "Oh yeah, you need to run the etcd test suite to check that this namespace works". ;-)

File: jepsen/src/jepsen/generator/interpreter.clj (hunk 1 above)
Comment (stevana) [reply jepsen/src/jepsen/generator/interpreter.clj:49]: (This was caught by the linter btw.)

File: jepsen/test/jepsen/util_test.clj [hunk 2]
```diff
@@ -78,11 +77,11 @@
 
 (deftest letr-test
   (testing "no bindings"
-    (is (= (letr []) nil))
-    (is (= (letr [] 1 2) 2)))
```
Comment (aphyr) [inline jepsen/test/jepsen/util_test.clj:82]: This is another case where I feel like the linter is being overzealous. We *want* this intermediate form to go unused! That's what we're testing for. :-)

File: jepsen/test/jepsen/util_test.clj [hunk 3]
```diff
@@ -112,14 +111,14 @@
 
 (deftest timeout-test
   ; Fast operations pass through the inner result or exception.
-  (is ::success (timeout 1000 ::timed-out
-                         ::success))
+  (is (= ::success (timeout 1000 ::timed-out
```
Comment (aphyr) [inline jepsen/test/jepsen/util_test.clj:114]: Oy. That's been broken a while then, hasn't it.

File: jepsen/src/jepsen/checker/perf.clj [hunk 4]
```diff
@@ -489,7 +489,7 @@
         history     (util/history->latencies history)
         datasets    (invokes-by-f-type history)
         fs          (util/polysort (keys datasets))
-        fs->points  (fs->points fs)
+        fs->points' (fs->points fs)
```
[the remark this answers is not in the document]
Comment (stevana) [reply jepsen/src/jepsen/checker/perf.clj]: Fixed.

File: jepsen/src/jepsen/cli.clj [hunk 5]
```diff
@@ -329,7 +329,9 @@ Options:\n")
                    (web/serve! options)
                    (info (str "Listening on http://"
                               (:ip options) ":" (:port options) "/"))
-                   (while true (Thread/sleep 1000)))}})
```
[the remark this answers is not in the document]
Comment (stevana) [reply jepsen/src/jepsen/cli.clj:332]: True, the linter looks for constant true or false conditionals. I haven't found a way to disable warnings on a per function basis...

File: jepsen/src/jepsen/store.clj [hunk 6]
```diff
@@ -312,7 +313,7 @@
           dest (.. FileSystems
                    getDefault
                    (getPath base-dir (into-array dest)))]
-      (Files/deleteIfExists dest)
+      (util/discard-ret-val (Files/deleteIfExists dest))
```
[the remark this answers is not in the document]
Comment (stevana) [reply jepsen/src/jepsen/store.clj:316]: This lint rule looks for unused return values. It was the same rule that found the bug with `set` vs `set!`, so I don't think disabling it is a great idea.

The problem here is that `Files/deleteIfExists` returns a boolean indicating if it deleted the file or not, we don't care about that boolean and I think it makes sense to signal that in the code. The `util/discard-ret-val` functional is a mouthful, perhaps we can come up with something nicer? In Haskell land we got `Control.Monad.void : Functor f => f a -> f ()` which throws away the return value. Or perhaps find a delete file function that doesn't return a boolean?

File: jepsen/src/jepsen/os/centos.clj [hunk 7]
```diff
@@ -44,7 +44,7 @@
          (update!))))
 
 (defmacro dprint [x]
-  `((fn [x#] do
+  `((fn [x#] ; XXX: 'Could not resolve var: do
```
[the remark this answers is not in the document]
Comment (stevana) [reply jepsen/src/jepsen/os/centos.clj:47]: Fixed.

File: jepsen/src/jepsen/generator/interpreter.clj [hunk 8]
```diff
@@ -231,7 +231,7 @@
                     (do (doseq [[thread queue] invocations]
                           (.put ^ArrayBlockingQueue queue {:type :exit}))
                         ; Wait for exit
-                        (mapv (comp deref :future) workers)
+                        (util/discard-ret-val (mapv (comp deref :future) workers))
```
[the remark this answers is not in the document]
Comment (stevana) [reply jepsen/src/jepsen/generator/interpreter.clj]: `doseq (map ...` doesn't, but `dorun (map ...` does. Fixed.

File: jepsen/test/jepsen/util_test.clj (hunk 2 above)
Comment (stevana) [reply jepsen/test/jepsen/util_test.clj:82]: I'll try a bit more to disable lint rules on a per function basis.

File: jepsen/src/jepsen/cli.clj [hunk 9]
```diff
@@ -390,8 +392,14 @@ Options:\n")
...
-                            (assert+ stored-test IllegalStateException
-                                     "Not sure what the last test was")
+                            ;; XXX: constant-test: Test expression is always logical
+                            ;; true or always logical false: (clojure.core/map?
+                            ;; m__8038__auto__) in form (if (clojure.core/map?
+                            ;; m__8038__auto__) (clojure.core/ex-info "Assert
+                            ;; failed" m__8038__auto__) (new
+                            ;; IllegalStateException m__8038__auto__))
+                            ;; (assert+ stored-test IllegalStateException
+                            ;;          "Not sure what the last test was")
```
[the remark this answers is not in the document]
Comment (stevana) [reply jepsen/src/jepsen/cli.clj]: I'm still confused about how to fix this, but I did manage to disable it -- see `.eastwood.clj`.

File: jepsen/test/jepsen/util_test.clj (hunk 2 above)
Comment (stevana) [reply jepsen/test/jepsen/util_test.clj:82]: Fixed.

File: jepsen/src/jepsen/store.clj (hunk 6 above)
[the remark this answers is not in the document]
Comment (stevana) [reply jepsen/src/jepsen/store.clj:316]: I moved the `Files/deleteIfExists` to the let binding above it, where I could discard the value. It's perhaps not the most elegant solution, but with this change we can remove `util/discard-ret-val` from the code base.

**Comment by stevana** [comment]
>  "Oh yeah, you need to run the etcd test suite to check that this namespace works". ;-)

Btw, how do you test an unpublished snapshot? Do you build a local snapshot and use that, or do you point it to the repo and use git commit hash? The former seems a bit clunky, and when I've looked into the latter I couldn't find a solution that seemed maintained, solid and easy to use?

File: jepsen/src/jepsen/cli.clj (hunk 9 above)
[the remark this answers is not in the document]
Comment (aphyr) [reply jepsen/src/jepsen/cli.clj]: It's cuz assert+ macroexpands into something like (if (map? {...}) (throw+ ...)), I think

File: jepsen/src/jepsen/store.clj (hunk 6 above)
[the remark this answers is not in the document]
Comment (aphyr) [reply jepsen/src/jepsen/store.clj]: Blergh. All right, let's go with that. I still feel like this is weirdly persnickety--side effects are a thing, and I'm a believer in linters not complaining about idiomatic code, but if that's the only case we have to change...

**Comment by aphyr** [comment]
> Btw, how do you test an unpublished snapshot? Do you build a local snapshot and use that, or do you point it to the repo and use git commit hash? The former seems a bit clunky, and when I've looked into the latter I couldn't find a solution that seemed maintained, solid and easy to use?

Lein checkouts! I keep a `checkouts` directory in my test dirs--e.g. `jepsen.etcd/checkouts`, with a symlink to `jepsen/jepsen`. As long as the project.clj versions align, it should run the code from the checkouts dir, rather than the Maven cache. Might have to run `lein install` in `jepsen/jepsen` once, just so lein picks up on the existence of the snapshot version.

**Comment by aphyr** [comment]
All right, let's merge it! Thanks Stevan! :)

**Comment by aphyr** [comment]
BTW, there's a big API change happening in Jepsen right now, and I think you're working on some changes to generators, right? Take a look at https://github.com/jepsen-io/jepsen/blob/master/jepsen/src/jepsen/generator/pure.clj for background and migration info. There's a `jepsen.generator/stateful+pure` wrapper that's providing backwards compatibility for the upcoming migration release, then jepsen.generator will be replaced by jepsen.generator.pure in 0.2.0.

**Comment by stevana** [comment]
Cool, looking forward to not seeing the interrupted sleep error anymore, :-).

What's the plan with regards to nemesis generators? They are all still currently stateful.

**Comment by stevana** [comment]
OK, I'll wait for your change before looking into reopening the clock nemesis targets PR then.

**Comment by aphyr** [comment]
They're live now, though... they may not actually *work* yet. That's what I've got to work on next. :)

**Pull request merged**